### ISSUE TITLE
Add country specific speeds

### ISSUE THREAD

**Issue by dnesbitt61** [valhalla/valhalla#1069]
Add logic to set speeds based on country. See: https://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Maxspeed#Default_speed_limits

This also refers to places= tags - so issue #236 is related.

https://wiki.openstreetmap.org/wiki/Places

**Comment by westnordost** [comment]
I have been researching into this and was about to open a ticket here, asking how Valhalla infers implicit speed limits (= when no sign is posted).

In StreetComplete, I am [working on sourcing the information](https://github.com/westnordost/StreetComplete/issues/813) that is missing to infer these speed l limits now also for the United States, as the jurisdiction differs for each state. The tags set will then probably look something like i.e.
`maxspeed:type=US-AK:residential` (= road uses the implicit speed limit for residential districts in Alaska)

The linked wiki page is far from complete. It would be nice, if there was an implementation for that data. This doesn't need to be part of Valhalla but could also be another library that is then simply used by Valhalla. 
90% of this library would be researching, determining and maintenance of these implicit speed limits - that is, rule sets written in per-country configuration files.
If anyone is poised to do this - casting the legislations that determine implicit speed limits per state into machine-readable configuration files, I might give some input there as I both already researched into the topic and have some ideas how these files could be structured.

**Comment by 1ec5** [comment]
Note that, although @westnordost’s [distillation of default limits](https://wiki.openstreetmap.org/wiki/Default_speed_limits) is much simplified from the source legislation, there are a lot of distinctions that require context beyond just the containing national and regional administrative boundaries. For example, a common feature among U.S. state rules is that “residential” areas are treated differently than “business” districts – a zoning rule that might even be out of scope for OSM. (Other rules are a little more tractable, taking into consideration the highway class, route network, or number of lanes.)

So while I think the basic ability to vary the default speed limit by national or regional boundary is useful, we should keep in mind the limits of this approach for automatically, authoritatively backfilling for missing speed limits or missing speed limit signs.

**Comment by ianthetechie** [comment]
Agreed. At the moment, I see this not as a way to backfill data into OSM directly, but to provide more sane defaults for the purposes of estimating route duration.

**Comment by westnordost** [comment]
For cases like residence and business zones, indeed, it is an open question how to record these. Same with the urban /rural distinction in most places outside the US as well.

We can distinguish two cases:

1. The speed limit for the road type or zone mentioned in the legislation is always signed. For example, to my knowledge, the speed limit of school zones in the US are always signed. Then, there ain't a problem because the "default speed limit" rules are overridden by signed ones anyway.

2. The speed limit for that road type is not explicitly signed. In that case the road type must be tagged somehow. For urban, the current tag scheme for that is maxspeed:type=DE:urban (or source:maxspeed as key) for example.
In the future hopefully a more reasonable tagging can be found. This part requires a our several proposals and the goal of the research I did was actually to find which tags would be necessary.

In the legislations I researched, it is often not clear whether (some of) the stated rules are intended for motor vehicle drivers to know by heart or if they are intended as rules for the road traffic authority where to put the speed limit sign. 

This is exactly the part that is missing from my research because it requires the help of people having local knowledge.

Am 10. Mai 2019 23:09:31 MESZ schrieb "Minh Nguyễn" <notifications@github.com>:

**Comment by westnordost** [comment]
For cases like residence and business zones, indeed, it is an open question how to record these. Same with the urban /rural distinction in most places outside the US as well.


We can distinguish two cases:


1. The speed limit for the road type or zone mentioned in the legislation is always signed. For example, to my knowledge, the speed limit of school zones in the US are always signed. Then, there ain't a problem because the "default speed limit" rules are overridden by signed ones anyway.


2. The speed limit for that road type is not explicitly signed. In that case the road type must be tagged somehow. For urban, the current tag scheme for that is maxspeed:type=DE:urban (or source:maxspeed as key) for example.
In the future hopefully a more reasonable tagging can be found. This part requires a our several proposals and the goal of the research I did was actually to find which tags would be necessary.


In the legislations I researched, it is often not clear whether (some of) the stated rules are intended for motor vehicle drivers to know by heart or if they are intended as rules for the road traffic authority where to put the speed limit sign. 


This is exactly the part that is missing from my research because it requires the help of people having local knowledge.

To list these missing pieces is something that parser can help too (cause I imagine it would emit an error or warning there)

**Comment by westnordost** [comment]
To avoid confusion, the default speed limits table can be used for two purposes:
Not only to guess a speed limit when no explicit one is recorded but also and primarily so to determine the speed limit on the absence of a sign, so when no explicit maxspeed will ever be recorded.

**Comment by ianthetechie** [comment]
What are your thoughts on this as a first pass at making the wiki table machine readable? https://github.com/stadiamaps/osm-default-speeds/blob/master/default_speeds.json

I have to say it was a bit trickier than I expected given the extensive use of `colspan` and `rowspan` (I did try parsing wikitext directly as well, but the html parsing libs are much more mature). Standardization / mapping to OSM classes may be tricky, so I expect that we'll need to take a few passes at this.

**Comment by kevinkreiser** [comment]
@gknisely can you have a look at this, do the admin codes map to those we have?

**Comment by gknisely** [comment]
I was going to implement this in the following manner; however, I never seem to get the time to do it.  Maybe this will help...

First, the admin db has 2 tables.  
- `admins` contains the following info `admin_level|iso_code|parent_admin|name|name_en|drive_on_right|geom'
- `admin_access` contains the following info `admin_id|iso_code|trunk|trunk_link|track|footway|pedestrian|bridleway|cycleway|path|motorroad
`

`Admins` contains administrative info for countries(level 2) and state/providence(level 4).  
 `Admin_access` contains access overrides for particular countries.  The `admin_id` is a foreign key to the `rowid` in Admins.  

```spatialite> select * from admin_access;
admin_id|iso_code|trunk|trunk_link|track|footway|pedestrian|bridleway|cycleway|path|motorroad
303|US|||||262|262|262|774|
132|GB||||||262|262||
64|CH|1225|1225|||||516||
1156|ES||||||||774|1229
126|RU|||||||516||
219|RO|1225|1225||||||774|
54|PL|1225|1225|||||||
2152|NO||||262|262||262||
1921|NL|1225|1225|||||262|774|
8|HU|1225|1225|||||||
6|SK|1225|1225|||||||
1205|FR|1225|1225|||262|||774|
99|FI|||||262||262||
58|DK|1225|1225|262||||262||
123|BR||||774||774|774|774|
385|TR||||||258|||
78|BE|1225|1225|||262|258|774|774|
117|BY||||262|262||262||
7|AT|1225|1225||||||258|
199|AU||||262|262|262|262||
```
And we can look up the country via: 

```
spatialite> select * from admins where rowid = 1921;
admin_level|iso_code|parent_admin|name|name_en|drive_on_right|geom
2|NL||Nederland|The Netherlands|1|
```
So, I was going to implement a speed table with the same type of logic: foreign key, country speed, urban speed, rural speed, etc. (etc., could mean default speeds for each highway tag in that country or better yet a mask.  Not sure.) I was going to hopefully parse this data from OSM via the [admin.lua file](https://github.com/valhalla/valhalla/blob/master/lua/admin.lua).

I was going to implement the speed override logic in the [enhancer](https://github.com/valhalla/valhalla/blob/master/src/mjolnir/graphenhancer.cc#L1269) because we have the density info.  This was going to be used to determine if we were in a rural vs urban area.  
Logic would be as follows:

1. Determine if we are in a rural or urban area.
2. If we have a speed based on the area from # 1, then update the speed; otherwise, update the speed if we have a country speed or highway tag speeds for that country. 
3. Otherwise, do nothing.  

That is a lot of info.  I hope this helps.

**Comment by westnordost** [comment]
@ianthetechie I answered you in the tracker for the conversion script:
https://github.com/stadiamaps/osm-default-speeds/issues

Since it is made for valhalla guys, perhaps they could give some input as well on the desired format, for example here: https://github.com/stadiamaps/osm-default-speeds/issues/3

**Comment by 1ec5** [comment]
This discussion has continued in #3334.

**Comment by nilsnolde** [comment]
I wonder if we shouldn't go directly for https://github.com/westnordost/osm-legal-default-speeds as discussed in #3334.. It seems it's "just" parsing the OSM wiki page turning it into a JSON. I'm not sure yet what'd be easiest, try to use it natively as a library (I fear a rabbit hole) or write our own access functions. We'd really only need a subset. In the end I didn't look at it yet too much, but since it's based on the OSM wiki page, this seems superior to maintaining a mapping ourselves, which would become stale pretty quickly. I recently saw Graphhopper implemented it as well. But for them it's simpler with kotlin -> java.

Especially for other modes than car (e.g. HGV), this would help a ton.
